src/algebra/is_smul_regular.lean

```diff
@@ -0,0 +1,182 @@
+  rw [← nat.succ_pred_eq_of_pos n0, pow_succ', ← smul_eq_mul],
+  exact is_smul_regular.of_smul _,
+end
+
+end monoid
+
+section monoid_with_zero
+
+variables [monoid_with_zero R] [monoid_with_zero S] [has_zero M] [mul_action_with_zero R M]
+  [mul_action_with_zero R S] [mul_action_with_zero S M] [is_scalar_tower R S M]
+
+/--  The element `0` is `M`-regular if and only if `M` is trivial. -/
+lemma is_smul_regular.subsingleton (h : is_smul_regular M (0 : R)) : subsingleton M :=
+⟨λ a b, h (by repeat { rw mul_action_with_zero.zero_smul })⟩
+
+/--  The element `0` is `M`-regular if and only if `M` is trivial. -/
+lemma is_smul_regular_zero_iff_subsingleton : is_smul_regular M (0 : R) ↔ subsingleton M :=
+⟨λ h, h.subsingleton, λ H a b h, @subsingleton.elim _ H a b⟩
+
+/--  The `0` element is not `M`-regular, on a non-trivial semimodule. -/
+lemma not_is_smul_regular_zero_iff : ¬ is_smul_regular M (0 : R) ↔ nontrivial M :=
+begin
+  rw [nontrivial_iff, not_iff_comm, is_smul_regular_zero_iff_subsingleton, subsingleton_iff],
+  push_neg,
+  exact iff.rfl
+end
+
+/--  The element `0` is `M`-regular when `M` is trivial. -/
+lemma is_smul_regular_zero [sM : subsingleton M] : is_smul_regular M (0 : R) :=
+is_smul_regular_zero_iff_subsingleton.mpr sM
+
+/--  The `0` element is not `M`-regular, on a non-trivial semimodule. -/
+lemma not_is_smul_regular_zero [nM : nontrivial M] : ¬ is_smul_regular M (0 : R) :=
+not_is_smul_regular_zero_iff.mpr nM
+
+/-- An element of `S` admitting a left inverse in `R` is `M`-regular. -/
+lemma is_smul_regular_of_smul_eq_one (h : a • s = 1) : is_smul_regular M s :=
+is_smul_regular.of_smul a (by { rw h, exact is_smul_regular_one })
+
+/-- An element of `R` admitting a left inverse is `M`-regular. -/
+lemma is_smul_regular_of_mul_eq_one (h : a * b = 1) : is_smul_regular M b :=
+is_smul_regular.of_smul a (by { rw [smul_eq_mul, h], exact is_smul_regular_one })
+
+/-- Any element in `units R` is `M`-regular. -/
+lemma units.is_smul_regular (a : units R) : is_smul_regular M (a : R) :=
+begin
+  have ai : a.inv • a.val = 1,
+  { rw smul_eq_mul,
+     exact a.inv_val },
+  { exact is_smul_regular_of_smul_eq_one ai }
+end
+
+/-- A unit is `M`-regular. -/
+lemma is_unit.is_smul_regular (ua : is_unit a) : is_smul_regular M a :=
+begin
+  rcases ua with ⟨a, rfl⟩,
+  exact units.is_smul_regular a,
+end
+
+end monoid_with_zero
+
+section comm_monoid
+open module
+
+variables [comm_monoid R] [mul_action R M]
+
+/--  A product is `M`-regular if and only if the factors are. -/
+lemma is_smul_regular_mul_iff : is_smul_regular M (a * b) ↔
+  is_smul_regular M a ∧ is_smul_regular M b :=
+begin
+  rw [← is_smul_regular_mul_and_mul_iff],
+  exact ⟨λ ab, ⟨ab, by rwa [mul_comm]⟩, λ rab, rab.1⟩
+end
+
+end comm_monoid
+
+/-- Left-regularity in a `semiring R` is equivalent to `M`-regularity, when
+the `R`-module `M` is `R`.  -/
+lemma is_left_regular_iff {R : Type*} [semiring R] (a : R) :
+  is_left_regular a ↔ module.is_smul_regular R a :=
+iff.rfl
```

Comment:
Does this actually require `semiring`? Or is `monoid R` enough?

Reply:
Monoid is enough: changed!